### PR TITLE
🏗 Enable coverage reports on e2e testing

### PR DESCRIPTION
**Pull request by rcebulko**
When running `gulp e2e --coverage`, after each test completes, coverage data will be reported to the endpoint provided by `istanbul-middleware` which aggregates the coverage across all tests. Before the test server shuts down, it then downloads the aggregated coverage report and opens the coverage explorer UI.

Builds on #30004 and #30111. Won't work properly until #30099 is also merged, but won't break anything either (both because nothing is using the `--coverage` flag and because the test workflow is not dependent on the server endpoint), so it is safe to review and submit independently. PR #30128 contains the merged version of these two sets of changes, so you can check out the branch on that PR if you'd like to test/experiment locally.

Because some of this PR involved reorganizing `e2e/index.js`, it will be easiest to follow/review commit-by-commit 

> Note: For this to work, the compiled code must be instrumented. This happens automatically by default, but if using the `--nobuild` flag, you must be sure to also pass `--coverage` to `gulp build/dist` for instrumentation.

### CODE REVIEW

File: build-system/tasks/e2e/index.js
Comment (rcebulko) [inline build-system/tasks/e2e/index.js]: Fixed in a later commit

**Comment by rcebulko** [comment]
@rsimha @estherkim friendly ping

File: build-system/tasks/e2e/describes-e2e.js
Comment (rsimha) [inline build-system/tasks/e2e/describes-e2e.js:476]: `afterEach()` is called after every single test. Is the intent to report coverage once per test? If not, perhaps call `reportCoverage()` after all tests have completed, but before the server is stopped?

File: build-system/tasks/e2e/index.js
Comment (rsimha) [inline build-system/tasks/e2e/index.js:49]: Style nit: Can we convert these function-level comments to JSDoc style?

**Review by rsimha — COMMENTED** [review]
Thanks for working on this. I've left one high-level comment below re: when to report coverage. I'll defer to @estherkim for full review and approval since she is more familiar with the E2E code.

File: build-system/tasks/e2e/describes-e2e.js
Comment (rsimha) [reply build-system/tasks/e2e/describes-e2e.js:476]: I re-read the description and just noticed the bit about aggregating coverage. This is no longer a concern.

Comment (estherkim) [reply build-system/tasks/e2e/describes-e2e.js:476]: Unresolving this thread because making an XHR after each test is a concern of mine as well.

**Review by estherkim — COMMENTED** [review]
A concern is that coverage logic is sprinkled everywhere. Is there a reason why e2e coverage works differently than unit and integration tests?

It would be nice if all 3 worked similarly, i.e. create a coverage reporter and use a separate gulp task to upload results - if that's possible.

**Comment by rcebulko** [comment]
> A concern is that coverage logic is sprinkled everywhere. Is there a reason why e2e coverage works differently than unit and integration tests?

I wouldn't say it's sprinkled everywhere; there's logic in `e2e/index` to enable coverage and fetch the report, as well as `describes-e2e` containing the `afterEach` logic to report the coverage during a test This is only 2 places where coverage logic lives, each of which seems pretty reasonable.

Unit and integration tests are run in one location, so the coverage data itself can be collected locally. For e2e tests, the tests themselves are running in the browser, and there's no 

> It would be nice if all 3 worked similarly, i.e. create a coverage reporter and use a separate gulp task to upload results - if that's possible.

I agree this would be preferable. This is in a stage of active exploration, and most pieces involved are still subject to change. Part of the issue is that for every step (transpilation, sourcemapping, testing, reporting coverage, aggregating coverage) there are 2~4 different packages available that purport to do that job. The end goal here will be to have a similar format, where the report is uploaded separately by a dedicated gulp task (or, possibly, the coverage results might be combined with unit/integration tests for overall coverage).

> Unresolving this thread because making an XHR after each test is a concern of mine as well.

A goal of this PR for now is to make it possible to generate this "layer" of steps, so I can explore things like other/custom reporters (or even toy with CircleCI's coverage reporting) and collect info on what will work best. In parallel, I plan to modify `init_tests` to not clear the coverage variable globally, so that the coverage info can be reported one single time at the end of the full test suite. It's unclear how much that will entail though (including accounting for what happens when tests fail), so this PR is the first step here, with optimizations/improvements coming down the pipe soon. This is not enabled in any CI nor is it documented, but is a stepping stone to progressing further and fleshing out what the final picture should eventually look like.

I'd prefer to get this in with the understanding that it's experimental and will never be enabled in builds while the XHR-per-test is still in place. It will allow progress on other layers of the stack in parallel, and still makes it possible for members of other WGs (esp. UI) to try this out locally and provide feedback on how their team might make use of it.

File: build-system/tasks/e2e/describes-e2e.js
Comment (rcebulko) [reply build-system/tasks/e2e/describes-e2e.js:476]: @estherkim Addressed in comment below, PTAL

File: build-system/tasks/e2e/index.js
Comment (rcebulko) [reply build-system/tasks/e2e/index.js:49]: Done in local branch. `git push` doesn't seem to like using my phone hotspot for some reason, so it won't show up here until I get on a different source of internet :(

**Comment by estherkim** [comment]
Sorry for the delay, I wanted to make sure I understood your comments. Sounds more complicated than I anticipated.

Since this is experimental, would it be okay to see if a new or existing mocha reporter is feasible before going this route? I remember there was some exploration recently as part of the test-cases project.

**Comment by rcebulko** [comment]
> Sorry for the delay, I wanted to make sure I understood your comments. Sounds more complicated than I anticipated.
No worries! Appreciate you taking a closer look

> Since this is experimental, would it be okay to see if a new or existing mocha reporter is feasible before going this route? I remember there was some exploration recently as part of the test-cases project.
I believe @rsimha has tried that approach, and I've explored it somewhat as well. The main challenge is that Karma's built-in support for coverage instruments code on-the-fly, reporting coverage info based on the lines/branches/functions of the *compiled* code, not the original source. A custom Mocha reporter unfortunately can't access the in-page coverage object, as the Runner instance provided to the event handlers doesn't have get a reference to the page controller environment needed to extract the coverage.

File: build-system/tasks/e2e/describes-e2e.js
Comment (estherkim) [reply build-system/tasks/e2e/describes-e2e.js:475]: Even though this work is experimental as you say, I think this part is worth iterating over. How about writing to a local file?

Comment (rcebulko) [reply build-system/tasks/e2e/describes-e2e.js:475]: What are you suggesting we write to a local file?

If you're referring to the coverage data itself, the issue isn't just storing it, it's actually merging coverage objects across multiple runs. This functionality is built into the `istanbul-middleware` package that handles the coverage UI. I could presumably reverse-engineer that logic (there's an attempt at that elsewhere, though it's not maintained), or there may be a way to get into the plugin logic itself. I would prefer not to embed that attempt & complexity in this PR, instead committing this to iterate on as a base point.

Comment (rcebulko) [reply build-system/tasks/e2e/describes-e2e.js:475]: @estherkim @rsimha I'd really like to resolve this this week if possible so I can make progress in adjacent areas without the code changing underneath

Comment (estherkim) [reply build-system/tasks/e2e/describes-e2e.js:475]: Totally fair, maybe @rsimha can weigh in since you and I seem to be at an impasse on what's acceptable as experimental :)

Comment (rcebulko) [reply build-system/tasks/e2e/describes-e2e.js:475]: To add color, e2e tests each make multiple HTTP requests _to and from the browser_, while this adds one request _that takes place within Node, not the browser_. There is no apparent impact on runtime with vs. without this coverage reporting.

File: build-system/tasks/e2e/index.js
Comment (estherkim) [reply build-system/tasks/e2e/index.js:13]: Did you push the commit?

File: build-system/tasks/e2e/describes-e2e.js
Comment (estherkim) [reply build-system/tasks/e2e/describes-e2e.js:475]: Good point, okay then. Approving this to unblock.

**Review by estherkim — APPROVED** [review]

File: build-system/tasks/e2e/describes-e2e.js
Comment (rcebulko) [reply build-system/tasks/e2e/describes-e2e.js:475]: Thanks :+1:

Comment (rsimha) [reply build-system/tasks/e2e/describes-e2e.js:475]: Just caught up with the comments. Thanks for iterating on this @rcebulko and @estherkim!

Here are the three highlights that stood out to me:

> 1. _"I'd prefer to get this in with the understanding that it's experimental and will never be enabled in builds while the XHR-per-test is still in place."_
> 2. "_In parallel, I plan to modify init_tests to not clear the coverage variable globally, so that the coverage info can be reported one single time at the end of the full test suite._"
> 3. "_There is no apparent impact on runtime with vs. without this coverage reporting._"

Based on these three facts, I'm comfortable with the current code being checked in.

**Pull request merged**
